### PR TITLE
drivers: dma: add shim driver for mcux edma

### PR DESCRIPTION
**Pull request by jfischer-no**
This patch restructures Kconfig (de41725) and adds support for the mcux edma controller.
The shim driver have base support for the three transport variants, but may need a few improvements (also Kconfig.mcux), WIP/RFC.

### CODE REVIEW

File: drivers/dma/Kconfig [hunk 1]
```diff
@@ -101,22 +51,23 @@ config DMA_2_IRQ_PRI
 	help
 	IRQ Priority for DMA Controller 2.
 
-config DMA_2_RX_SUB_CHANNEL_ID
-	int "Requested RX sub-channel ID"
+config SYS_LOG_DMA_LEVEL
```
Comment (galak) [inline drivers/dma/Kconfig]: Don't introduce new Kconfig options in a refactor patch

File: drivers/dma/Kconfig.mcux [hunk 2]
```diff
@@ -0,0 +1,90 @@
+# Kconfig - MCUXpresso SDK DMA
+#
+#
+#
+# Copyright (c) 2017, Phytec Messtechnik GmbH
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig DMA_MCUX
```
Comment (MaureenHelm) [inline drivers/dma/Kconfig.mcux]: Reorder your commits so the Kconfig gets introduced before the shim, or just squash the Kconfig into the shim commit

File: tests/drivers/dma/test_loop_transfer/prj_frdm_k64f.conf [hunk 3]
```diff
@@ -0,0 +1,4 @@
+CONFIG_DMA=y
+CONFIG_DMA_MCUX=y
```
Comment (MaureenHelm) [inline tests/drivers/dma/test_loop_transfer/prj_frdm_k64f.conf]: Instead of this, let's add a dma section in Kconfig.defconfig.mk64f12. Can you also update the testcase.ini file so sanitycheck builds this test for frdm_k64f?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/dma/edma_mcux.c [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+	/* TODO: support for the period trigger ? */
+	/*DMAMUX_EnablePeriodTrigger(DMAMUX0, channel);*/
+
+	DMAMUX_EnableChannel(DMAMUX0, channel);
+
+	EDMA_CreateHandle(&data[channel].handle, base, channel);
+	EDMA_SetCallback(&data[channel].handle, edma_mcux_callback, dev);
+	/* TODO: necessary ? */
+	/*EDMA_ResetChannel(base, channel); ? */
```
Comment (jfischer-phytec-iot) [inline drivers/dma/edma_mcux.c]: I'll check it later

File: drivers/dma/edma_mcux.c [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+	return 0;
+}
+
+static int dma_mcux_stop(struct device *dev, u32_t channel)
+{
+	struct dma_mcux_data *data = dev->driver_data;
+
+	SYS_LOG_DBG("");
+
+	EDMA_StopTransfer(&data[channel].handle);
```
Comment (jfischer-phytec-iot) [inline drivers/dma/edma_mcux.c]: ok

File: drivers/dma/edma_mcux.c [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+static int dma_mcux_configure(struct device *dev)
+{
+	const struct dma_mcux_config *config = dev->config->config_info;
+	struct dma_mcux_data *data = dev->driver_data;
+	DMA_Type *base = config->base;
+	edma_config_t edma_config;
+
+	SYS_LOG_DBG("");
+
+	data[0].dma_callback = NULL;
```
Comment (jfischer-phytec-iot) [inline drivers/dma/edma_mcux.c]: Should not be there, forgotten to remove, as I see it is not necessary to initialize it with zeros.

File: drivers/dma/edma_mcux.c [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+	DMAMUX_Init(DMAMUX0);
+
+	EDMA_GetDefaultConfig(&edma_config);
+
+	edma_config.enableContinuousLinkMode = false;
+	edma_config.enableRoundRobinArbitration = false;
+	edma_config.enableDebugMode = true;
```
Comment (jfischer-phytec-iot) [inline drivers/dma/edma_mcux.c]: ok

File: drivers/dma/Kconfig.mcux (hunk 2 above)
Comment (jfischer-phytec-iot) [inline drivers/dma/Kconfig.mcux]: ok

File: tests/drivers/dma/test_loop_transfer/prj_frdm_k64f.conf (hunk 3 above)
Comment (jfischer-phytec-iot) [inline tests/drivers/dma/test_loop_transfer/prj_frdm_k64f.conf]: ok

File: arch/arm/soc/nxp_kinetis/k6x/Kconfig.defconfig.mk64f12 [hunk 8]
```diff
@@ -81,4 +81,11 @@ config UART_MCUX
 
 endif # SERIAL
 
+if DMA
+
+config DMA_MCUX
+	def_bool y
```
Comment (MaureenHelm) [inline arch/arm/soc/nxp_kinetis/k6x/Kconfig.defconfig.mk64f12]: Please do this in a separate commit. Can also enable for KW41

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Go ahead and actually squash the two -squash- commits, then force push.

You can remove the gerrit commit-msg hook, we don't need the Change-Id anymore.

File: drivers/dma/Kconfig.mcux [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+endchoice
+
+config DMA_MCUX_NUMOF_DMA_CHANNELS
+	int "Number of enabled DMA channels"
+	default 2
+	range 1 8
+	help
+	  This option configures the number of enabled DMA channels.
+
+config DMA_MCUX_TCD0_QUEUE_SIZE
```
Comment (galak) [inline drivers/dma/Kconfig.mcux]: Is it reasonable to have a TCD queue of size of 0 for a used TCD?

File: ext/hal/nxp/mcux/drivers/Makefile [hunk 10]
```diff
@@ -14,3 +14,5 @@ obj-$(CONFIG_SPI_MCUX_DSPI) += fsl_dspi.o
 obj-$(CONFIG_UART_MCUX) += fsl_uart.o
 obj-$(CONFIG_UART_MCUX_LPSCI) += fsl_lpsci.o
 obj-$(CONFIG_UART_MCUX_LPUART) += fsl_lpuart.o
+obj-$(CONFIG_DMA_MCUX) += fsl_edma.o
+obj-$(CONFIG_DMA_MCUX) += fsl_dmamux.o
```
Comment (galak) [inline ext/hal/nxp/mcux/drivers/Makefile]: just put them on one line:

obj-$(CONFIG_DMA_MCUX) += fsl_edma.o fsl_dmamux.o

**Review by galak — CHANGES_REQUESTED** [review]
Would like to see DTS support for this before we accepted it.

**Review by galak — COMMENTED** [review]
I'm pulling out the commits that are generic and not related to MCUX dma and push a PR for them (#1062)

**Comment by jfischer-phytec-iot** [comment]
>Would like to see DTS support for this before we accepted it.

I'll make it, but it can take a little.

File: drivers/dma/Kconfig.mcux (hunk 9 above)
Comment (jfischer-phytec-iot) [inline drivers/dma/Kconfig.mcux]: It is then disabled.

**Comment by jfischer-phytec-iot** [comment]
@galak I have started with DTS support, please look at it, if it goes in the right direction.

**Comment by galak** [comment]
Sorry, didn't noticed you updated this.

Have you looked at:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/dma/dma.txt

We should conform to that for general dma concepts, like channels, etc.

**Comment by jfischer-phytec-iot** [comment]
>We should conform to that for general dma concepts, like channels, etc.

@galak yes, I have already reworked it but not pushed to this PR, will do after ELCE is over

**Comment by jfischer-phytec-iot** [comment]
@galak Can we merge it without DT support?

**Comment by galak** [comment]
> @galak Can we merge it without DT support?

Nope.

**Comment by jfischer-phytec-iot** [comment]
That's unfair, it was not necessary for the Atmel SAM DMA driver...

**Comment by nashif** [comment]
what is the story with this PR? Is this going to be updated and is it still required? If nothing happens within the next week, it will be closed.

**Comment by nashif** [comment]
@jfischer-phytec-iot ping, can this be closed?

**Comment by jfischer-phytec-iot** [comment]
@nashif I would like to revise it again
